RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc

```diff
@@ -164,6 +164,400 @@ namespace lowptgsfeleid {
+        TVector3 gsfTV3(0, 0, 0);
+        gsfTV3.SetPtEtaPhi(gsf->ptMode(), gsf->etaMode(), gsf->phiMode());
+        TVector3 eleTV3(0, 0, 0);
+        eleTV3.SetPtEtaPhi(ele->pt(), ele->eta(), ele->phi());
+        gsf_dr = eleTV3.DeltaR(gsfTV3);
+      }
+    }
+
+    // Super clusters
+    if (ele->core().isNonnull()) {
+      reco::SuperClusterRef sc = ele->core()->superCluster();
+      if (sc.isNonnull()) {
+        eid_sc_E = sc->energy();
+        eid_sc_eta = sc->eta();
+        eid_sc_etaWidth = sc->etaWidth();
+        eid_sc_phiWidth = sc->phiWidth();
+        sc_Nclus = sc->clustersSize();
+      }
+    }
+
+    // Track-cluster matching
+    if (ele.isNonnull()) {
+      eid_match_seed_dEta = ele->deltaEtaSeedClusterTrackAtCalo();
+      eid_match_eclu_EoverP = (1. / ele->ecalEnergy()) - (1. / ele->p());
+      eid_match_SC_EoverP = ele->eSuperClusterOverP();
+      eid_match_SC_dEta = ele->deltaEtaSuperClusterTrackAtVtx();
+      eid_match_SC_dPhi = ele->deltaPhiSuperClusterTrackAtVtx();
+    }
+
+    // Shower shape vars
+    if (ele.isNonnull()) {
+      eid_shape_full5x5_HoverE = ele->full5x5_hcalOverEcal();
+      eid_shape_full5x5_r9 = ele->full5x5_r9();
+    }
+
+    // Misc
+    eid_rho = rho;
+
+    if (ele.isNonnull()) {
+      eid_brem_frac = ele->fbrem();
+      core_shFracHits = ele->shFracInnerHits();
+    }
+
+    // Unbiased BDT from ElectronSeed
+    gsf_bdtout1 = unbiased;
+
+    // Clusters
+    if (ele->core().isNonnull()) {
+      reco::GsfTrackRef gsf = ele->core()->gsfTrack();
+      if (gsf.isNonnull()) {
+        reco::SuperClusterRef sc = ele->core()->superCluster();
+        if (sc.isNonnull()) {
+          // Propagate electron track to ECAL surface
+          double mass_ = 0.000511 * 0.000511;
+          float p2 = pow(gsf->p(), 2);
+          float energy = sqrt(mass_ + p2);
+          XYZTLorentzVector mom = XYZTLorentzVector(gsf->px(), gsf->py(), gsf->pz(), energy);
+          XYZTLorentzVector pos = XYZTLorentzVector(gsf->vx(), gsf->vy(), gsf->vz(), 0.);
+          float field_z = 3.8;
+          RawParticle particle(mom, pos);
+          BaseParticlePropagator propagator(particle, 0., 0., field_z);
+          particle.setCharge(gsf->charge());
+          propagator.propagateToEcalEntrance(true);   // true only first half loop , false more than one loop
+          bool reach_ECAL = propagator.getSuccess();  // 0 does not reach ECAL, 1 yes barrel, 2 yes endcaps
+          GlobalPoint ecal_pos(particle.x(),          // ECAL entry point for track
+                               particle.y(),
+                               particle.z());
+
+          // Iterate through ECAL clusters and sort in energy
+          int clusNum = 0;
+          float maxEne1 = -1;
+          float maxEne2 = -1;
+          int i1 = -1;
+          int i2 = -1;
+          try {
+            if (sc->clustersSize() > 0 && sc->clustersBegin() != sc->clustersEnd()) {
+              for (auto& cluster : sc->clusters()) {
```

Comment:
```suggestion
              for (auto const& cluster : sc->clusters()) {
```
same for other range loops

Reply:
not addressed

Reply:
Done.